### PR TITLE
[prowl] Initial contribution

### PR DESCRIPTION
**Pull request by octa22**
This is the binding for the Prowl (https://www.prowlapp.com) iOS push service.  
It has been written from scratch and therefore it is not based on the original 1.x Prowl binding.  
It has no other purpose than sending push messages to iOS devices.

Signed-off-by: Ondrej Pecta <opecta@gmail.com>

### CODE REVIEW

File: bundles/org.openhab.binding.prowl/pom.xml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.2.0-SNAPSHOT</version>
```
Comment (wborn) [inline bundles/org.openhab.binding.prowl/pom.xml:10]:
```suggestion
    <version>3.3.0-SNAPSHOT</version>
```

File: bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlBindingConstants.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
```
Comment (wborn) [inline bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlBindingConstants.java:2]:
```suggestion
 * Copyright (c) 2010-2022 Contributors to the openHAB project
```

File: bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
```
Comment (wborn) [inline bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlConfiguration.java:2]:
```suggestion
 * Copyright (c) 2010-2022 Contributors to the openHAB project
```

File: bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
```
Comment (wborn) [inline bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java:2]:
```suggestion
 * Copyright (c) 2010-2022 Contributors to the openHAB project
```

File: bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(ProwlConfiguration.class);
+        updateStatus(ThingStatus.UNKNOWN);
+
+        statusFuture = scheduler.scheduleWithFixedDelay(() -> {
+            updateStatus();
+        }, 0, config.refresh, TimeUnit.MINUTES);
```
Comment (wborn) [inline bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java:75]:
```suggestion
        statusFuture = scheduler.scheduleWithFixedDelay(() -> updateStatus(), 0, config.refresh, TimeUnit.MINUTES);
```

File: bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+                    .content(
+                            new StringContentProvider("apikey=" + config.apiKey + "&application=" + config.application
+                                    + "&event=" + event + "&description=" + description),
+                            "application/x-www-form-urlencoded; charset=UTF-8")
+                    .send();
+            String resp = response.getContentAsString();
+            updateFreeMessages(resp);
+            if (logger.isTraceEnabled()) {
+                logger.trace("add response: {}", resp);
+            }
```
Comment (wborn) [inline bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java:141]: The logger only does string substitution if the log level matches. So it should have not much impact on performance.

```suggestion
            logger.trace("add response: {}", resp);
```

File: bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandlerFactory.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
```
Comment (wborn) [inline bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandlerFactory.java:2]:
```suggestion
 * Copyright (c) 2010-2022 Contributors to the openHAB project
```

**Review by wborn — COMMENTED** [review]
Here are some comments you might want to incorporate when you continue working on this. :+1: 

It also seems the binding is still missing from the BOM file in `bom/openhab-addons/pom.xml` and you can also add yourself to the CODEOWNERS file. :slightly_smiling_face:

File: bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/action/ProwlActions.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
+package org.openhab.binding.prowl.internal.action;
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/action/ProwlActions.java:13]: added header

File: bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandlerFactory.java (hunk 7 above)
Comment (octa22) [reply bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandlerFactory.java]: fixed

File: bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java (hunk 6 above)
Comment (octa22) [reply bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java]: fixed

File: bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+        super.dispose();
+    }
+
+    private boolean keyVerificationSucceeded(String apiKey) {
+        try {
+            ContentResponse response = httpClient.GET(PROWL_VERIFY_URI + "?apikey=" + apiKey);
+            String resp = response.getContentAsString();
+            if (logger.isTraceEnabled()) {
+                logger.trace("verify response: {}", resp);
+            }
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java]: fixed

File: bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java (hunk 5 above)
Comment (octa22) [reply bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java]: fixed

File: bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java (hunk 4 above)
Comment (octa22) [reply bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlHandler.java]: Fixed

File: bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlBindingConstants.java (hunk 2 above)
Comment (octa22) [reply bundles/org.openhab.binding.prowl/src/main/java/org/openhab/binding/prowl/internal/ProwlBindingConstants.java]: fixed

File: bundles/org.openhab.binding.prowl/pom.xml (hunk 1 above)
Comment (octa22) [reply bundles/org.openhab.binding.prowl/pom.xml]: rebased and fixed

**Comment by wborn** [comment]
Thanks for addressing my comments @octa22! It seems a somfytahoma commit now also ended up on this branch. Can you remove it and also make SAT happy by adding a new line at the end of the readme?

**Comment by octa22** [comment]
Thanks @wborn , I hope that everything is OK now.

**Comment by wborn** [comment]
Can you also add the default translations to `/OH-INF/i18n/prowl.properties` so the texts can be translated using Crowdin? You can generate these using: `mvn i18n:generate-default-translations`

It is also possible to make the ThingActions texts translatable using `@text/` identifiers. See for instance [PushoverActions](https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/actions/PushoverActions.java) and [pushover.properties](https://github.com/openhab/openhab-addons/blob/9e6b901d539cdae7cff586d8795df293c5d85a51/bundles/org.openhab.binding.pushover/src/main/resources/OH-INF/i18n/pushover.properties#L41-L84).

**Review by wborn — APPROVED** [review]
Thanks a lot! :+1: Let's merge it now everything has been addressed. :rocket:

**Comment by jamesfawcett** [comment]
Hi @octa22 

Would it be possible to get priority added to the prowl notifications?

For example in the OH1 bindings, we could use this format:

`pushNotification("{key}", "Battery Alert - Battery low on z-wave device.\n\n ", **5** )`

We had to pass the key again as the first input, then the message, and we could send priority 0 - 5. Each different number makes a different notification sound on an iOS device. It would be great if you knew a way for the new bindings to support this too?

**Pull request merged**